Automatically manage engine versions

pchote opened the pull request:
Closes #7.

I would like to get a second/third set of eyes on this before merging – especially the Linux/macOS scripts, which I have reworked from the first version to follow the more sensible approach of @abcdefg30's powershell scripts.

Phrohdoh commented on launch-dedicated.sh:
```diff
@@ -23,7 +28,12 @@ ADVERTISE_ONLINE="${AdvertiseOnline:-"True"}"
-cd engine
+cd "${TEMPLATE_ROOT}"
+if [ ! -d "${ENGINE_DIRECTORY}" ]; then
+	make -f Makefile
```
Why are you doing this?

make will see makefiles in the current directory and use that before the makefile in the engine.

pchote replied:
That's the idea, yeah.  If the engine doesn't exist it runs make *on the template* to compile the mod which then fetches the engine.

Phrohdoh replied:
That is the default make behavior.

pchote replied:
Ah, now I get you.  You are referring to the `-f Makefile`, right?  that was a leftover from an older version where I was `-f ${TEMPLATE_ROOT}/Makefile`ing.

pchote commented:
Updated.  Thinking about the makefile situation a bit more, I came to the conclusion that automatically fetching and building *everything* is unexpected and a bit rude.  I have changed the launch scripts to instead prompt the user to run this step manually.

abcdefg30 commented on launch-game.cmd:
```diff
@@ -8,11 +8,22 @@ if %INCLUDE_DEFAULT_MODS% neq "True" goto launch
+echo Required engine files not found.
+echo Run `make all` in the mod directory to fetch and build the required files, then try again.
+pause
 exit /b
```
You should `cd %TEMPLATE_DIR%` here as well.

pchote replied:
Fixed this and the other dedicated/utility scripts by instead not changing into the engine directory in this case.

pchote commented:
Updated.

pchote commented:
I am planning to merge this tomorrow evening if there are no more issues.  Will be travelling friday/saturday and plan to tackle #13 and #8 while offline, which both depend on this being finished.

reaperrr commented:
I don't know if it's just on my end, but when running 'make all', I get a 'CommandNotFoundException' for `Expand-Archive`.

pchote commented:
Looks like that command requires a newer version of powershell than you have.  I will replace it with a different extraction technique.

pchote commented:
Updated.  Please try again @reaperrr.

reaperrr commented:
> Updated. Please try again @reaperrr.

Works now, the only remaining thing:
It first says `Version set to foo`, but then in the next line `unknown command 'version'`. The example mod launches fine, but after closing the example mod via Quit the command line windows throws an error, because mod.yaml - Metadata.Version is still {DEV_VERSION}.
```
Failed to register currrent mod metadata
System.ArgumentException: Ein Element mit dem gleichen Schlüssel wurde bereits hinzugefügt.
   bei System.ThrowHelper.ThrowArgumentException(ExceptionResource resource)
   bei System.Collections.Generic.Dictionary'2.Insert(TKey key, TValue value, Boolean add)
   bei System.Collections.Generic.Dictionary'2.Add(TKey key, TValue value)
   bei OpenRA.ExternalMods.LoadMod(MiniYaml yaml) in e:\ORA_ModTemplate1\engine\OpenRA.Game\ExternalMods.cs:Zeile 83.
   bei OpenRA.ExternalMods.Register(Manifest mod) in e:\ORA_ModTemplate1\engine\OpenRA.Game\ExternalMods.cs:Zeile 115.
```

abcdefg30 commented:
> It first says `Version set to foo`, but then in the next line `unknown command 'version'`.

This is probably the bug in the engine's make script and fixed by https://github.com/OpenRA/OpenRA/pull/13359.

pchote commented:
> Failed to register currrent mod metadata
System.ArgumentException:

This is another engine bug that is fixed by https://github.com/OpenRA/OpenRA/pull/13367

reaperrr commented:
Tested again with fetching latest bleed, bug is gone 👍 
I can't help with the Linux/macOS scripts, but on the Windows side things look mergable (maybe update last commit to fetch latest bleed before merging).

pchote commented:
Done.
